src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,317 @@
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
+                return;
+            }
+
+            foreach (var disposeScope in DisposeScopeStack) {
+                bool wasRemoved = disposeScope.Disposables.Remove(disposable);
+                if (wasRemoved && !disposeScope.DontReportOnExternalDisposes) {
+                    // File.AppendAllText(@"c:\slask\disp.txt",
+                    //     "\n\nWho did it:"+
+                    //     disposeScope.StackTrace.ToString().Substring(0, 1000));
+                }
+            }
+        }
+
+        internal static DisposeScope NewDisposeScope()
+        {
+            lock (Singleton ??= new DisposeScopeManager()) {
+                return Singleton.InnerNewDisposeScope();
+            }
+        }
+
+        private DisposeScope InnerNewDisposeScope()
+        {
+            var disposeScope = new DisposeScope(this);
+            DisposeScopeStack.Push(disposeScope);
+            return disposeScope;
+        }
+
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                if (DisposeScopeStack.Count == 0) {
+                    throw new InvalidOperationException($"There are no DisposeScopes! ");
+                }
+
+                if (DisposeScopeStack.Peek() != disposeScope) {
+                    throw new InvalidOperationException(
+                        $"DisposeScope for {disposeScope.Disposables} disposed out of order!");
+                }
+
+                DisposeScopeStack.Pop();
+            }
+        }
+
+        internal void MoveToOuterScope(DisposeScope disposeScope, IDisposable disposable)
+        {
+            lock (this) {
+                var array = DisposeScopeStack.ToArray();
+                for (var i = array.Length - 1; i >= 1; i--) {
+                    if (array[i] == disposeScope) {
+                        array[i - 1].Include(disposable);
+                        return;
+                    }
+                }
+            }
+        }
+
+        public class DisposeScope : IDisposable
+        {
+            private readonly DisposeScopeManager _disposeScopeManager;
+
+            public DisposeScope(DisposeScopeManager disposeScopeManager)
+            {
+                _disposeScopeManager = disposeScopeManager;
+                // StackTrace = new StackTrace();
+            }
+
+            // public StackTrace StackTrace { get; set; }
+
+            /// <summary>
+            /// The disposables that are scheduled for disposing.
+            /// </summary>
+            public HashSet<IDisposable> Disposables { get; } = new();
+
+            public long TotalAllocatedSize =>
+                Disposables.OfType<torch.Tensor>().Sum(x => x.NumberOfElements * x.ElementSize);
+
+            public bool DontReportOnExternalDisposes { get; set; } = true;
```

Comment:
I wonder about this one -- it seems to me that external disposes are the norm, and not in anyway an error. The DisposeScope should simply ignore tensors that have already been disposed.

Reply:
I added that for some debugging purposes, it allowed me to trap when the disposes were happening. All disposes were appropriate, it turned out.

Reply:
I removed it, if we need that kind of debugging, it's easy enough to re-introduce.